flux-middle/src/rty/mod.rs

```diff
@@ -373,6 +391,54 @@ impl AdtDef {
     }
 }
 
+impl TypeFoldable for Qualifier {
+    fn super_fold_with<F: TypeFolder>(&self, folder: &mut F) -> Self {
+        Qualifier {
+            name: self.name.clone(),
+            args: self.args.clone(),
+            expr: self.expr.fold_with(folder),
+        }
+    }
+
+    fn super_visit_with<V: TypeVisitor>(&self, visitor: &mut V) {
+        self.expr.visit_with(visitor)
+    }
+}
+
+impl TypeFoldable for AdtDef {
+    fn super_fold_with<F: TypeFolder>(&self, folder: &mut F) -> Self {
+        AdtDef(Interned::new(AdtDefData {
+            def_id: self.def_id(),
+            sorts: self.sorts().to_vec(),
+            flags: *self.flags(),
+            nvariants: self.0.nvariants,
+            opaque: self.0.opaque,
+            invariants: self
+                .invariants()
+                .iter()
+                .map(|inv| inv.fold_with(folder))
+                .collect_vec(),
+        }))
+    }
+
+    fn super_visit_with<V: TypeVisitor>(&self, visitor: &mut V) {
+        self.invariants()
+            .iter()
+            .for_each(|inv| inv.visit_with(visitor));
+    }
+}
+
+impl TypeFoldable for Invariant {
+    fn super_fold_with<F: TypeFolder>(&self, folder: &mut F) -> Self {
+        let pred = self.pred.fold_with(folder);
+        Invariant { pred }
+    }
+
+    fn super_visit_with<V: TypeVisitor>(&self, visitor: &mut V) {
+        self.pred.visit_with(visitor);
+    }
+}
```

Comment:
These implementations should go in the `fold` module for consistency

Reply:
done!